File: Codepath/Week5-StacksAndQueues/java/QueueUsingStacks.java
/*
 * Problem: Implement a queue using instances of stack data structure and operations on them.
*/
class MyQueue {
    private Stack<Integer> s1 = new Stack<>();
    private Stack<Integer> s2 = new Stack<>();
    private int front;

    /** Push element x to the back of queue. */
    public void enQueue(int x) {
         if (s1.empty())
            front = x;
        s1.push(x);
    }

    /** Removes the element from in front of queue
     * and returns that element. */
    public int deQueue() {
        if (s2.isEmpty()) {
            while (!s1.isEmpty())
                s2.push(s1.pop());
        }
        return s2.pop();
    }

    public int peek() {
        if (!s2.isEmpty()) {
            return s2.peek();
        }
        return front;
    }

    /** Returns whether the queue is empty. */
    public boolean empty() {
        return s1.isEmpty() && s2.isEmpty();
    }

File: Codepath/Week5-StacksAndQueues/python/QueueUsingStacks.py
# Problem: Implement a queue using instances of stack data structure and operations on them.
class MyQueue:

    s1 = []
    s2 = []
    front = -1

    def enQueue(self, x):
        if (len(self.s1) == 0):
            front = x
        self.s1.append(x)

    def deQueue(self):
        if len(self.s2) == 0:
            while(len(self.s1) != 0):
                self.s2.append(self.s1.pop())
        return self.s2.pop()

    def peek(self):
        if len(self.s2) != 0:
            return self.s2[0]
        return self.front

    def empty(self):
        return len(self.s2) == 0 and len(self.s1) == 0
